src/watchdog.js

```diff
@@ -0,0 +1,401 @@
+
+		return Promise.resolve()
+			.then( () => this.destroy() )
+			.then( () => {
+				if ( typeof this._elementOrData === 'string' ) {
+					return this.create( this._data, this._config );
+				}
+
+				const updatedConfig = Object.assign( {}, this._config, {
+					initialData: this._data
+				} );
+
+				return this.create( this._elementOrData, updatedConfig );
+			} )
+			.then( () => {
+				this.fire( 'restart' );
+			} );
+	}
+
+	/**
+	 * Creates a watched editor instance using the creator passed to the {@link #setCreator} method.
+	 *
+	 * @param {HTMLElement|String} elementOrData
+	 * @param {module:core/editor/editorconfig~EditorConfig} [config]
+	 *
+	 * @returns {Promise.<module:watchdog/watchdog~Watchdog>}
+	 */
+	create( elementOrData, config ) {
+		if ( !this._creator ) {
+			/**
+			 * @error watchdog-creator-not-defined
+			 *
+			 * The watchdog creator is not defined, define it using `watchdog.setCreator()`.
+			 */
+			throw new CKEditorError(
+				'watchdog-creator-not-defined: The watchdog creator is not defined, define it using `watchdog.setCreator()`.',
+				null
+			);
+		}
+
+		if ( !this._destructor ) {
+			/**
+			 * @error watchdog-destructor-not-defined
+			 *
+			 * The watchdog destructor is not defined, define it using `watchdog.setDestructor()`.
+			 */
+			throw new CKEditorError(
+				'watchdog-destructor-not-defined: The watchdog destructor is not defined, define it using `watchdog.setDestructor()`',
+				null
+			);
+		}
+
+		this._elementOrData = elementOrData;
+		this._config = config;
+
+		return Promise.resolve()
+			.then( () => this._creator( elementOrData, config ) )
+			.then( editor => {
+				this._editor = editor;
+
+				window.addEventListener( 'error', this._boundErrorHandler );
+				this.listenTo( editor.model.document, 'change:data', this._throttledSave );
+
+				this._lastDocumentVersion = editor.model.document.version;
+				this._data = editor.getData();
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Destroys the current editor using the destructor passed to {@link #setDestructor} method.
+	 *
+	 * @returns {Promise.<module:watchdog/watchdog~Watchdog>}
+	 */
+	destroy() {
+		window.removeEventListener( 'error', this._boundErrorHandler );
+		this.stopListening( this._editor.model.document, 'change:data', this._throttledSave );
+
+		return Promise.resolve()
+			.then( () => this._destructor( this._editor ) )
+			.then( () => {
+				this._editor = null;
+
+				return this;
+			} );
+	}
+
+	/**
+	 * Saves the editor data, so it can be restored after the crash even if the data cannot be fetched at
+	 * the moment of a crash.
+	 *
+	 * @private
+	 */
+	_save() {
+		const version = this._editor.model.document.version;
+
+		// Change may not produce an operation, so the document's version
+		// can be the same after that change.
+		if ( version < this._lastDocumentVersion ) {
+			this._throttledSave.cancel();
+
+			return;
+		}
+
+		try {
+			this._data = this._editor.getData();
```

Comment:
```suggestion
			this._data = this._editor.data.get();
```

Using `data.get` might be safer, since theoretically `getData` is not required in the editor interface.